scipy/stats/stats.py

```diff
@@ -5117,24 +5346,105 @@ def ks_2samp(data1, data2):
     (0.07999999999999996, 0.41126949729859719)
 
     """
+    LARGE_N = 10000  # 'auto' will attempt to be exact if n1,n2 <= LARGE_N
     data1 = np.sort(data1)
     data2 = np.sort(data2)
     n1 = data1.shape[0]
     n2 = data2.shape[0]
+    if min(n1, n2) == 0:
+        raise ValueError('Data passed to ks_2samp must not be empty')
+
     data_all = np.concatenate([data1, data2])
+    # using searchsorted solves equal data problem
     cdf1 = np.searchsorted(data1, data_all, side='right') / n1
     cdf2 = np.searchsorted(data2, data_all, side='right') / n2
-    d = np.max(np.absolute(cdf1 - cdf2))
-    # Note: d absolute not signed distance
-    en = np.sqrt(n1 * n2 / (n1 + n2))
-    try:
-        prob = distributions.kstwobign.sf((en + 0.12 + 0.11 / en) * d)
-    except Exception:
-        warnings.warn('This should not happen! Please open an issue at '
-                    'https://github.com/scipy/scipy/issues and provide the code '
-                    'you used to trigger this warning.\n')
-        prob = 1.0
+    cddiffs = cdf1 - cdf2
+    minS = -np.min(cddiffs)
+    maxS = np.max(cddiffs)
+    alt2Dvalue = {'less': minS, 'greater': maxS, 'two-sided': max(minS, maxS)}
+    d = alt2Dvalue[alternative]
+    g = gcd(n1, n2)
+    n1g = n1 // g
+    n2g = n2 // g
+    prob = -np.inf
+    original_mode = mode
+    if mode == 'auto':
+        if max(n1, n2) <= LARGE_N:
+            mode = 'exact'
+        else:
+            mode = 'asymp'
+    elif mode == 'exact':
+        # If lcm(n1, n2) is too big, switch from exact to asymp
+        if n1g >= np.iinfo(np.int).max / n2g:
```

Comment:
This `if` is never true for tests

Reply:
This one is a little time- and memory-consuming to trigger as it requires constructing two arrays each with about 3 billion elements.